Failed on AbDb file: 1qd0_0H

**Describe the bug**
`cdrclu` failed on `1qd0_0H`

**To Reproduce**
Steps to reproduce the behavior:
1. Running environment
  - Version used: 0.1.1
  - Operating system and version: Ubuntu 22.04
2. Input data: AbDb `1qd0_0H`
3. Running commands:
```sh
$ cd tmp 
$ cdrclu -db /AbDb -o tmp -c L3 1qd0_0H
```
4. Error
```plaintext 
[01/02/24 14:59:25] WARNING  1qd0_0H chain chain_type='H' not exist ...                                                                                                                                                                  app.py:171
14:59:25 {/home/vscode/.conda/envs/abag_analysis/lib/python3.11/site-packages/cdrclass/examine_abdb_struct.py:282} [ERROR] examine_abdb_struct - pdb1qd0_0H chain_type='L' does not exist ... FAILED [MainThread]
                    WARNING  1qd0_0H chain chain_type='L' not exist ...                                                                                                                                                                  app.py:171
Traceback (most recent call last):
  File "/home/vscode/.conda/envs/abag_analysis/bin/cdrclu", line 8, in <module>
    sys.exit(app())
             ^^^^^
  File "/home/vscode/.conda/envs/abag_analysis/lib/python3.11/site-packages/cdrclass/app.py", line 930, in app
    main(args=args)
  File "/home/vscode/.conda/envs/abag_analysis/lib/python3.11/site-packages/cdrclass/app.py", line 906, in main
    results, metadata = worker(
                        ^^^^^^^
  File "/home/vscode/.conda/envs/abag_analysis/lib/python3.11/site-packages/cdrclass/app.py", line 779, in worker
    criteria, struct_df, metadata = process_single_mar_file(
                                    ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.conda/envs/abag_analysis/lib/python3.11/site-packages/cdrclass/app.py", line 176, in process_single_mar_file
    criteria["chain_length_okay"] = assert_seqres_atmseq_length(struct_fp=struct_fp,
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vscode/.conda/envs/abag_analysis/lib/python3.11/site-packages/cdrclass/examine_abdb_struct.py", line 356, in assert_seqres_atmseq_length
    assert len(seq) >= len(atmseq[cid])
                           ~~~~~~^^^^^
KeyError: 'A'
```

**Expected behavior**
NA

**Screenshots**
NA

**Additional context**
NA

--- a/.github/pull_request_template.md
+++ b/.github/pull_request_template.md
@@ -1,0 +1,35 @@
+# Pull Request Template
+
+## Description
+Please include a summary of the change and which issue is fixed. Also, include relevant motivation and context. List any dependencies that are required for this change.
+
+Fixes # (issue number)
+
+## Type of change
+
+Please mark the options that are relevant.
+
+- [ ] Bug fix (non-breaking change which fixes an issue)
+- [ ] New feature (non-breaking change which adds functionality)
+- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
+- [ ] This change requires a documentation update
+
+## How Has This Been Tested?
+
+Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.
+
+- [ ] Test A
+- [ ] Test B
+
+## Checklist:
+
+- [ ] I have performed a self-review of my own code
+- [ ] I have commented my code, particularly in hard-to-understand areas
+- [ ] I have made corresponding changes to the documentation
+- [ ] My changes generate no new warnings
+- [ ] I have added tests that prove my fix is effective or that my feature works
+- [ ] New and existing tests pass locally with my changes
+
+## Additional notes:
+
+Add any other details about the PR here.

--- a/cdrclass/abdb.py
+++ b/cdrclass/abdb.py
@@ -14,10 +14,7 @@ from Bio.PDB.Selection import unfold_entities
 from Bio.SeqIO import PdbIO
 
 # logging
-import logging
-logging.basicConfig(level=logging.INFO,
-                    format='%(asctime)s {%(pathname)s:%(lineno)d} [%(levelname)s] %(name)s - %(message)s [%(threadName)s]',
-                    datefmt='%H:%M:%S')
+from loguru import logger
 
 # custom 
 from cdrclass.pdb import chain2df
@@ -58,6 +55,74 @@ CDR_HASH_REV = dict(
 
 
 # ==================== Function ====================
+def summarize_ab_type_from_chain_map(chain_map: pd.DataFrame) -> List[str]:
+    """
+    Summarize antibody type from chain_map
+    E.g. chain_map 
+        chain_type  chain_label  chain_original
+    0            H            H               A
+
+    Args:
+        chain_map (pd.DataFrame): _description_
+
+    Returns:
+        List[str]: e.g.
+        ['single-domain', 'single-heavy'] or 
+        ['single-domain', 'single-light'] or 
+        ['double-domain', 'double-heavy'] or
+        ['double-domain', 'double-light'] or
+        ['double-domain', 'heavy-light']
+    """
+    ab_type = []
+    # count "H" and "L" in chain_type
+    num_ab_chains = 0
+    counts = chain_map["chain_type"].value_counts()
+    num_H, num_L = counts.get("H", 0), counts.get("L", 0)
+    num_ab_chains = sum([num_H, num_L])
+    if num_ab_chains == 1:
+        ab_type.append("single-domain")
+        if num_H == 1:
+            ab_type.append("single-heavy")
+        elif num_L == 1:
+            ab_type.append("single-light")
+    elif num_ab_chains == 2:
+        ab_type.append("double-domain")
+        if num_H == 2:
+            ab_type.append("double-heavy")
+        if num_L == 2:
+            ab_type.append("double-light")
+        if num_H == 1 and num_L == 1:
+            ab_type.append("heavy-light")
+    elif num_ab_chains > 2:
+        ab_type.append("multi-domain")
+    return ab_type
+
+
+def get_chain_map_from_remark_950(
+        abdb_fp: Path,
+        return_ab_type: bool = False
+) -> Tuple[float, List[str], pd.DataFrame]:
+    # parse REMARK 950
+    with open(abdb_fp, "r") as f:
+        remark950 = []
+        l = f.readline()
+        while not l.startswith("REMARK 950"):
+            l = f.readline()
+        while l.startswith("REMARK 950"):
+            remark950.append(l.strip())
+            l = f.readline()
+
+    # chain_map
+    cols = ["chain_type", "chain_label", "chain_original"]
+    chain_map = pd.DataFrame(
+        [l.split()[-3:] for l in remark950[2:]],
+        columns=cols
+    )
+    if return_ab_type:
+        return chain_map, summarize_ab_type_from_chain_map(chain_map)
+    return chain_map
+
+
 def get_resolution_from_abdb_file(abdb_struct_fp: Path) -> float:
     """ get resolution from .mar file """
     with open(abdb_struct_fp, "r") as f:
@@ -177,7 +242,7 @@ def extract_bb_atoms(struct_df: pd.DataFrame, include_CB: bool = True, add_resid
             try:
                 assert a in atms
             except AssertionError:
-                logging.warning(f"CAUTION: Atom type {a} not found in {c}{r}{alt}, resn: {resn}, cdr: {cdr}")
+                logger.warning(f"CAUTION: Atom type {a} not found in {c}{r}{alt}, resn: {resn}, cdr: {cdr}")
 
     # add residue identifier if True
     if add_residue_identifier:

--- a/cdrclass/app.py
+++ b/cdrclass/app.py
@@ -24,7 +24,7 @@ from cdrclass.examine_abdb_struct import (
     assert_struct_file_exist, 
     assert_seqres_atmseq_length,
     assert_no_cdr_missing_residues, 
-    assert_cdr_no_big_b_factor_core,
+    assert_cdr_no_big_b_factor,
     assert_cdr_no_non_proline_cis_peptide
 )
 from cdrclass.geometry import (
@@ -32,8 +32,8 @@ from cdrclass.geometry import (
     superimpose_atoms, atom_wise_dist
 )
 from cdrclass.utils import calc_omega_set_residues, calc_phi_psi_set_residues
-from cdrclass.exceptions import ClassifierNotExistError
-from cdrclass.abdb import CDR_HASH_REV, extract_bb_atoms, get_resolution_from_abdb_file
+from cdrclass.exceptions import *
+from cdrclass.abdb import CDR_HASH_REV, extract_bb_atoms, get_resolution_from_abdb_file, get_chain_map_from_remark_950
 
 # logger 
 from loguru import logger
@@ -162,7 +162,13 @@ def process_single_mar_file(
     if ckpt_file_pass or not strict:
         # get atmseq and seqres
         atmseq, seqres = extract_atmseq_seqres(struct_fp=struct_fp)
-
+        
+        # get chain map from abdb file 
+        chain_map = get_chain_map_from_remark_950(abdb_fp=struct_fp)
+        ab_chain_labels = chain_map.query('chain_type in ["H", "L"]').chain_label.to_list()
+        atmseq = {k: v for k, v in atmseq.items() if k in ab_chain_labels}
+        seqres = {k: v for k, v in seqres.items() if k in ab_chain_labels}
+        
         # 1. check chain_types exist 
         metadata['chain_type_exists'] = {}
         for chain_type in chain_types:
@@ -196,30 +202,24 @@ def process_single_mar_file(
             retain_b_factor=True,
             retain_hetatm=False,
             retain_water=False)
-        
-        criteria["cdr_no_missing_residue"] = True
-        for d in df_H + df_L:
-            b, details = assert_no_cdr_missing_residues(struct_fp=struct_fp,
-                                                        struct_df=d['df'],
-                                                        chain_type=d['chain_type'],
-                                                        chain_label=d['chain_label'],
-                                                        atmseq=atmseq[d['chain_label']], 
-                                                        seqres=seqres[d['chain_label']],
-                                                        clustal_omega_executable=clustal_omega_exe_fp,
-                                                        numbering_scheme=numbering_scheme)
-            metadata['cdr_missing_residue'][d['chain_label']] = {'pass': b, 'cdr_has_missing_residue (T->missing; F->no missing)': details}
-            criteria["cdr_no_missing_residue"] = criteria["cdr_no_missing_residue"] and b
-            if not b: 
-                # report details 
-                logger.warning(f"{abdbid} chain_label={d['chain_label']}, chain_type={d['chain_type']} has missing residues ...")
-                for k, v in details.items():
-                    s = 'has missing residues' if v else 'no missing residues'
-                    logger.warning(f"{abdbid} {k}: {s}")
+        criteria["cdr_no_missing_residue"] = all([
+            assert_no_cdr_missing_residues(struct_fp=struct_fp,
+                                                struct_df=d['df'],
+                                                chain_type=d['chain_type'],
+                                                chain_label=d['chain_label'],
+                                                atmseq=atmseq[d['chain_label']], 
+                                                seqres=seqres[d['chain_label']],
+                                                clustal_omega_executable=clustal_omega_exe_fp,
+                                                numbering_scheme=numbering_scheme)
+            for d in df_H + df_L
+        ])
+        if not criteria["cdr_no_missing_residue"]:
+            logger.warning(f"{abdbid} CDR ...")
 
         # 4. check loop CA B-factor (filter out ≥ 80 & == 0.)
         if criteria["cdr_no_missing_residue"]:
             criteria["cdr_no_big_b_factor"] = all([
-                assert_cdr_no_big_b_factor_core(struct_fp=struct_fp,
+                assert_cdr_no_big_b_factor(struct_fp=struct_fp,
                                                 struct_df=d['df'],
                                                 b_factor_atoms=b_factor_atom_set,
                                                 b_factor_thr=b_factor_cdr_thr,
@@ -262,7 +262,7 @@ def process_single_mar_file(
                 criteria["cdr_no_big_b_factor"], criteria["cdr_no_non_proline_cis_peptide"])):
             criteria["struct_okay"] = True
 
-    return criteria, struct_df, metadata
+    return criteria, struct_df
 
 # read cdr backbone atom coordinate csv
 def read_cdr_bb_csv(fp: Path=None, df: pd.DataFrame=None, add_residue_identifier: bool = False) -> pd.DataFrame:
@@ -548,6 +548,9 @@ def catch_error_from_process_cdr(func):
         except ClassifierNotExistError as e:
             logger.error(f"{func.__name__} raised an exception: {e}")
             return {'ClassifierNotExistError': str(e)}
+        except NaNDihedralError as e:
+            logger.error(f"{func.__name__} raised an exception: {e}")
+            return {'NaNDihedralError': str(e)}
     return wrapper
 
 # --------------------
@@ -580,10 +583,10 @@ def process_cdr(
     # ===== 0. retrieve pre-calculated classifiers =====
     clf_fps = list(classifier_dir.glob(f"{cdr}-{cdr_len}-*-FreeAb.joblib"))
     try:
-        assert len(clf_fps) > 0
+        assert clf_fps
     except AssertionError as e:
         raise ClassifierNotExistError(f"Cannot find classifier for {cdr=} and {cdr_len=}") from e
-    
+
     clfs = {
         fp.name.split("-FreeAb")[0]: joblib.load(filename=fp)
         for fp in clf_fps 
@@ -612,6 +615,11 @@ def process_cdr(
         angle_type=["phi", "psi"]
     )  # => shape (L, 4)
     X = loop_repr.reshape(1, -1)
+    # if np.nan in X, raise NaNDihedralError error 
+    if np.isnan(X).any():
+        logger.error(f"Found NaN dihedral angles in {abdbid=}, {cdr=}")
+        logger.error(f'Dihedral DataFrame:\n{dihedral_df}')
+        raise NaNDihedralError(f"{abdbid=}, {cdr=}, Found NaN dihedral angles")
 
     # ===== 3. find the closest LRC/CAN/AP to the query loop =====
     clf_name, dists, cluster_idx_list = [], [], []
@@ -753,7 +761,7 @@ def worker(
     cdr: str = None, 
     abdb_dir: Path = None, 
     classifier_dir: Path = None, 
-    lrc_ap_cluster: Dict[str, Any] = None) -> Tuple[List[Dict[str, Any]], Dict[str, Any]]:
+    lrc_ap_cluster: Dict[str, Any] = None) -> List[Dict[str, Any]]:
     """
     The core function that compare the input AbDb antibody with pre-calculated 
     Length and Residue Configuration (LRC) Affinity Propagation (AP) clusters
@@ -776,13 +784,13 @@ def worker(
     # ----------------------------------------
     # parse and validate the struct
     # ----------------------------------------
-    criteria, struct_df, metadata = process_single_mar_file(
+    criteria, struct_df = process_single_mar_file(
         struct_fp=abdb_dir.joinpath(f"pdb{abdbid}.mar"),
         abdbid=abdbid,
         strict=False,
     )
     if not criteria["cdr_no_missing_residue"]:
-        logger.warning(f"CDR {cdr} of {abdbid} has missing residue(s).")
+        logger.warning(f"{abdbid} CDR has missing residue(s).")
 
     # ----------------------------------------
     # extract dihedrals and backbone atoms
@@ -830,7 +838,7 @@ def worker(
                     })
             logger.info(f"Processing {abdbid}, CDR-{cdr} Done.")
             print("\n\n\n")
-    return results, metadata
+    return results
 
 # --------------------
 # Main
@@ -869,6 +877,9 @@ def cli() -> argparse.Namespace:  # sourcery skip: inline-immediately-returned-v
     # LRC AP cluster associated info file  
     parser.add_argument('-f', '--lrc_ap_info', type=Path, default=BASE/'assets'/'LRC_AP_cluster.json', 
                         help="LRC_AP_cluster.json file path that holds information about each precalculated LRC AP cluster")
+    # add a log file handle 
+    parser.add_argument('-l', '--log', type=Path, default=None, 
+                        help="log file path to save log info. If None, only print to stdout and stderr.")
     
     # parse args
     args = parser.parse_args()
@@ -877,6 +888,9 @@ def cli() -> argparse.Namespace:  # sourcery skip: inline-immediately-returned-v
 
 
 def main(args) -> None:
+    if args.log is not None:
+        logger.add(sink=args.log, level="DEBUG")
+        
     if is_first_run():
         # first run
         logger.info("First run, downloading assets ...")
@@ -903,7 +917,7 @@ def main(args) -> None:
     
     # block: run the worker 
     # ------------------------------------------------------------------------------
-    results, metadata = worker(
+    results: List[Dict[str, Any]] = worker(
         abdbid=abdbid, 
         cdr=cdr, 
         abdb_dir=abdb_dir, 
@@ -914,7 +928,6 @@ def main(args) -> None:
     
     # save results
     o = (outdir / f"{abdbid}.json").resolve()
-    results.append({'checkpoints': metadata})
     results.append({'job': {'abdbid': abdbid, 
                             'cdr': cdr,
                             'abdb_dir': str(abdb_dir),

--- a/cdrclass/examine_abdb_struct.py
+++ b/cdrclass/examine_abdb_struct.py
@@ -26,16 +26,14 @@ from typing import Dict, Tuple, Sequence, List, Optional, Union
 from Bio.PDB import PDBParser
 from Bio.PDB.Selection import unfold_entities
 from Bio.SeqIO import PdbIO
+# logging 
+from loguru import logger 
 
 # cdrclass 
 from cdrclass.pdb import chain2df
 from cdrclass.abdb import CDR_HASH_REV, CDR_HASH
 from cdrclass.run_align import pairwise_align_clustalomega, print_format_alignment
 from cdrclass.utils import calc_omega_set_residues
-
-# logging 
-import logging
-logger = logging.getLogger("examine_abdb_struct")
 
 
 def generate_anchor_residue_identifier(cdr_dict: Dict[str, Dict[str, List[str]]], numbering_scheme: str) -> Dict[str, List[Tuple[str, int]]]:
@@ -589,12 +587,12 @@ def assert_no_cdr_missing_residues(struct_fp: Path,
     return True, result
 
 
-def assert_cdr_no_big_b_factor_core(struct_fp: Path,
-                                    struct_df: pd.DataFrame,
-                                    b_factor_atoms: List[str],
-                                    b_factor_thr: float,
-                                    numbering_scheme: str,
-                                    ) -> bool:
+def assert_cdr_no_big_b_factor(struct_fp: Path,
+                               struct_df: pd.DataFrame,
+                               b_factor_atoms: List[str],
+                               b_factor_thr: float,
+                               numbering_scheme: str,
+                               ) -> bool:
     """
     Assert CDR loop specified atom set B factor is smaller than the thr by default 80.
     Args:

--- a/cdrclass/exceptions.py
+++ b/cdrclass/exceptions.py
@@ -5,3 +5,8 @@ class ClassifierNotExistError(Exception):
     def __init__(self, message):
         self.message = message
         super().__init__(self.message)
+
+class NaNDihedralError(Exception):
+    def __init__(self, message):
+        self.message = message
+        super().__init__(self.message)

--- a/cdrclass/utils.py
+++ b/cdrclass/utils.py
@@ -2,11 +2,7 @@ import numpy as np
 import pandas as pd
 from typing import List, Tuple
 # logging
-import logging
-logging.basicConfig(level=logging.INFO,
-                    format='%(asctime)s {%(pathname)s:%(lineno)d} [%(levelname)s] %(name)s - %(message)s [%(threadName)s]',
-                    datefmt='%H:%M:%S')
-logger = logging.getLogger(__name__)
+from loguru import logger 
 
 
 def calc_dihedral(
@@ -116,9 +112,12 @@ def calc_phi_psi_set_residues(
 
         # compute phi and psi angles for specified residue identifiers
         phi, psi = calc_phi_psi_single_residue(struct_df=struct_df, node_id=ni)
+        # convert to float if not None
+        phi = float(phi) if phi is not None else phi
+        psi = float(psi) if psi is not None else psi
 
         # append to result
-        phi_psi.append((ri, rn, float(phi), float(psi)))
+        phi_psi.append((ri, rn, phi, psi))
 
     return phi_psi
 

--- a/test/test_chain_map.py
+++ b/test/test_chain_map.py
@@ -1,0 +1,31 @@
+"""
+Use pytest to run this test.
+Test 1qd0_0H, whose chain mapping has non-protein chain A
+Should skip this chain in main processing 
+"""
+import pytest
+from pathlib import Path
+from loguru import logger
+from argparse import Namespace
+import cdrclass
+from cdrclass.app import main
+
+# ==================== Configuration ====================
+common_kwargs = {
+    'outdir': Path.cwd().joinpath('test_output'),
+    'abdb': Path('/AbDb'),
+    'lrc_ap_clusters': Path('/workspaces/CDRConformationClassification/dirs/classifier'),
+    'lrc_ap_info': Path('/workspaces/CDRConformationClassification/dirs/LRC_AP_cluster.json'),
+    'log': None
+}
+
+args = Namespace(abdbid='1qd0_0H', cdr='all', **common_kwargs)  # 1dcl_0: double-light
+
+# ==================== Tests ====================
+def test_main():
+    # Call the function and check if it runs without any errors
+    try:
+        print(f"Testing {args.abdbid}")
+        main(args)
+    except Exception as e:
+        pytest.fail(f"{args.abdbid}: function call failed with error: {e}")

--- a/test/test_nan_dihedral_angle_error.py
+++ b/test/test_nan_dihedral_angle_error.py
@@ -1,0 +1,30 @@
+"""
+Use pytest to run this test.
+Test 2h32_0P, which has a nan dihedral angle error.
+"""
+import pytest
+from pathlib import Path
+from loguru import logger
+from argparse import Namespace
+import cdrclass
+from cdrclass.app import main
+
+# ==================== Configuration ====================
+common_kwargs = {
+    'outdir': Path.cwd().joinpath('test_output'),
+    'abdb': Path('/AbDb'),
+    'lrc_ap_clusters': Path('/workspaces/CDRConformationClassification/dirs/classifier'),
+    'lrc_ap_info': Path('/workspaces/CDRConformationClassification/dirs/LRC_AP_cluster.json'),
+    'log': None
+}
+
+args = Namespace(abdbid='2h32_0P', cdr='all', **common_kwargs)  # 1dcl_0: double-light
+
+# ==================== Tests ====================
+def test_main():
+    # Call the function and check if it runs without any errors
+    try:
+        print(f"Testing {args.abdbid}")
+        main(args)
+    except Exception as e:
+        pytest.fail(f"{args.abdbid}: function call failed with error: {e}")

--- a/test/test_parse_double-same-domain_ab.py
+++ b/test/test_parse_double-same-domain_ab.py
@@ -17,6 +17,7 @@ common_kwargs = {
     'abdb': Path('/AbDb'),
     'lrc_ap_clusters': Path('/workspaces/CDRConformationClassification/dirs/classifier'),
     'lrc_ap_info': Path('/workspaces/CDRConformationClassification/dirs/LRC_AP_cluster.json'),
+    'log': None
 }
 
 # args = Namespace(abdbid='3b5g_0', cdr='all', **common_kwargs)  # 3b5g_0: double-light

--- a/test/test_parse_single-domain_ab.py
+++ b/test/test_parse_single-domain_ab.py
@@ -17,6 +17,7 @@ common_kwargs = {
     'abdb': Path('/AbDb'),
     'lrc_ap_clusters': Path(cdrclass.__path__[0]).joinpath('assets', 'classifier'),
     'lrc_ap_info': Path(cdrclass.__path__[0]).joinpath('assets', 'LRC_AP_cluster.json'),
+    'log': None
 }
 cases = [
     Namespace(abdbid='1u0q_1', cdr='all', **common_kwargs),  # 1u0q_1: single-heavy

